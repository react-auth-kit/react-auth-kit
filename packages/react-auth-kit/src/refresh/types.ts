/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Payload for Refresh token False
 */
type RefreshTokenActionResponseFailed = {
  /**
   * If the refresh operation is successful or not
   *
   * If the `isSuccess` is `true`, then the `token` and other items will be
   * replaced with the new network response
   *
   * If the isSuccess is `false`, then everything will be wiped and
   * the user will be signed out
   */
  isSuccess: false
}

/**
 * Payload for Refresh token Success True
 */
export type RefreshTokenActionResponsePassed<T> = {

  /**
   * New Auth token from the network response
   */
  newAuthToken: string,

  /**
   * New Auth Token type from the network response
   */
  newAuthTokenType?: string,

  /**
   * New Refresh token from the network response. Can be null
   */
  newRefreshToken?: string,

  /**
   * New User state from the network. Can be null
   */
  newAuthUserState?: T | null,

  /**
   * If the refresh operation is successful or not
   *
   * If the isSuccess is `true`, then the `token` and other items will be
   * replaced with the new network response
   *
   * If the isSuccess is `false`, then everything will be wiped and the user will
   * be signed out
   */
  isSuccess: boolean
}

/**
 * Response type by the refresh token
 */
export type RefreshTokenActionResponse<T> = RefreshTokenActionResponsePassed<T> | RefreshTokenActionResponseFailed;

/**
 * Callback function for the refresh token operation
 *
 * This function is called when the refresh token operation is triggered.
 * It should return a promise which resolves to a `RefreshTokenActionResponse`
 * object.
 *
 * @typeParam T - Type of User State Object
 */
export type refreshTokenCallback<T> = (param: {

  /**
   * Existing Auth token for the refresh operation
   */
  authToken?: string,

  /**
   * Existing Refresh token for the refresh operation
   */
  refreshToken?: string,

  /**
   * Existing User State for the User state
   */
  authUserState: T | null,
}) => Promise<RefreshTokenActionResponse<T>>
