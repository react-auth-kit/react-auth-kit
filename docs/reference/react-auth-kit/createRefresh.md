---
title: createRefresh - React Auth Kit
description: Create the Refresh Token engine for React Auth Kit
---


# createRefresh

<div data-ea-publisher="authkitarkadipme" data-ea-type="text" id="ref_createRefresh"></div>

## Import

```js
import createRefresh from 'react-auth-kit/createRefresh';
```


## Function Signature

**createRefresh**<`T`\>([`createRefreshParamInterface`](./types.md#createrefreshparaminterface)): [`createRefreshParamInterface`](./types.md#createrefreshparaminterface)<`T`\>

## Type Parameters

| Name | Description               |
|:-----|:--------------------------|
| `T`  | Type of User State Object |

## Parameters

| Name    | Type                                                                          | Required           | Description                                |
|:--------|:------------------------------------------------------------------------------|:-------------------|:-------------------------------------------|
| `param` | [`createRefreshParamInterface`](./types.md#createrefreshparaminterface)<`T`\> | :heavy_check_mark: | Parameters required for the refresh engine |

## Returns

[`createRefreshParamInterface`]()<`T`\>

Same params with an added layer of safety net.

## Example

```js
const refresh = createRefresh({
    interval: 10,
    refreshApiCallback: async (param) => {
        try {
            const response = await axios.post("/refresh", param, {
                headers: {'Authorization': `Bearer ${param.authToken}`}
            })
            console.log("Refreshing")
            return {
                isSuccess: true,
                newAuthToken: response.data.token,
                newAuthTokenExpireIn: 10,
                newRefreshTokenExpiresIn: 60
            }
        }
        catch(error){
            console.error(error)
            return {
                isSuccess: false
            }
        }
    }
})

```

## Remarks

This function doesn't really "do anything" at runtime, it just helps to organize the code base
Use this function to create the refresh token system

---

#### Defined in

[packages/react-auth-kit/src/createRefresh.ts:119](https://github.com/react-auth-kit/react-auth-kit/blob/37dc30d4/packages/react-auth-kit/src/createRefresh.ts#L119)

<p align="center">&mdash; ðŸ”‘  &mdash;</p>
<p align="center"><i>React Auth Kit is <a href="https://github.com/react-auth-kit/react-auth-kit/blob/master/LICENSE">MIT License</a> code</i></p>
