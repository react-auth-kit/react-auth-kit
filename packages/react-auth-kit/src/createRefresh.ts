/*
 * Copyright 2020 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {createRefreshParamInterface} from './types';

/**
 * This function doesn't really "do anything" at runtime,
 * it's just help to organize the code base
 * Use this function to create the refresh token system
 */
function createRefresh(param: createRefreshParamInterface)
  :createRefreshParamInterface {
  return param;
}

export default createRefresh;
