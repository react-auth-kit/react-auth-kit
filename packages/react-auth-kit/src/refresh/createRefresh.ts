/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use client';

import type React from 'react';

import {AuthKitConfigError} from "../error";
import {refreshTokenCallback} from "./types";


/**
 * Parameter for the Refresh operation
 */
export interface createRefreshAttribute<T> {

  /**
   * Interval on which the callback function is called
   */
  interval: number;

  /**
   * A Callback function which have the network request
   *
   * @example
   * ```js
   * refreshApiCallback: async (param) => {
   *  try {
   *    const response = await axios.post("/refresh", param, {
   *      headers: {'Authorization': `Bearer ${param.authToken}`}
   *    })
   *    console.log("Refreshing")
   *    return {
   *      isSuccess: true,
   *      newAuthToken: response.data.token,
   *      newAuthTokenExpireIn: 10,
   *      newRefreshTokenExpiresIn: 60
   *    }
   *  }
   *  catch(error){
   *    console.error(error)
   *    return {
   *      isSuccess: false
   *    }
   *  }
   * }
   * ```
   */
  refreshApiCallback: refreshTokenCallback<T>;

  /**
   * Initial component to be rendered when the refresh is not yet started
   * or the refresh is in progress.
   *
   * @remarks
   * This component is rendered only once when the refresh is started
   * and not at every refresh interval.
   * This is useful to show a loading spinner or a message
   *
   * @example
   * ```jsx
   * <Spinner />
   * ```
   */
  initialRefreshComponent?: React.ReactNode;
}

type createRefreshResponse<T> = createRefreshAttribute<T>

/**
 * @param param - Parameters required for the refresh engine
 * @returns Same params with added layer of a safety net.
 *
 * @remarks
 * This function doesn't really "do anything" at runtime,
 * it's just helps to organize the code base
 * Use this function to create the refresh token system
 *
 */
function createRefresh<T>(param: createRefreshAttribute<T>)
  :createRefreshResponse<T> {
  if (param.interval < 0) {
    throw new AuthKitConfigError(
        'Refresh interval is a time in seconds and can\'t be a negative(-ve)'+
        ' number. Make sure you are using positive number.',
    );
  }
  return param;
}

export default createRefresh;
