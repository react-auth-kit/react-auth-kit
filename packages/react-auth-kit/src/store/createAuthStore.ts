/*
 * Copyright 2020 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use client';

import TokenStore from "./TokenStore";
import {Store} from "./types";
import {AuthKitConfigError} from "../error";
import {createRefreshParamInterface} from "../refresh/createRefresh";
import {IStorage, cookieSameSite} from "../storage";
import CookieStorage from "../storage/CookieStorage";
import LocalStorage from "../storage/LocalStorage";
import DefaultStorageNamingStrategy from "../storage/DefaultStorageNamingStrategy";
import JwtToken from "../token/JwtToken";
import valueOrDefault from "../utils/valueOrDefault";

/**
 * Default Auth Types
 */
type defaultAuthTypes = 'localstorage' | 'cookie';

/**
 * Store Creation Param
 */
interface storeAttributes<T> {
  /**
   * The name of the cookie or local store object on which
   * the auth token is stored.
   *
   * This name will also be used as a prefix for all other cookies.
   */
  authName?: string;

  /**
   * Domain of the cookie, for which the cookie is valid.
   *
   * Needed if you are using `cookie` as authType
   *
   * @see {@link https://github.com/js-cookie/js-cookie#domain}
   */
  cookieDomain?: string;

  /**
   * Indicating if the cookie transmission requires a secure protocol (https).
   *
   * Needed if you are using `cookie` as authType
   *
   * @see {@link https://github.com/js-cookie/js-cookie#secure}
   */
  cookieSecure?: boolean;

  /**
   * Path of the cookie, for which the cookie is valid.
   *
   * Defaults to `/`
   *
   * Needed if you are using `cookie` as authType
   *
   * @see {@link https://github.com/js-cookie/js-cookie#path}
   */
  cookiePath?: string;

  /**
   * SameSite attribute for the cookie.
   *
   * Defaults to `none`
   *
   * Needed if you are using `cookie` as authType
   *
   * @see {@link https://github.com/js-cookie/js-cookie#samesite}
   */
  cookieSameSite?: cookieSameSite;

  /**
   * Refresh API. Created using the ` createRefresh ` function.
   */
  refresh?: createRefreshParamInterface<T>;

  /**
   * If Debug or not. Use this to debug your auth flow
   */
  debug?: boolean;
}

/**
 * Creates a new Auth Store for React Auth Kit.
 * This store can be used to manage authentication tokens and user state.
 * It supports both cookie and local storage as the underlying storage mechanism.
 *
 * @typeParam T - Type of User State Object
 *
 * @param authType - Type of the auth store. Can be either `cookie` or `localstorage`.
 * @param authAttributes - Attributes for the auth store, such as `authName`, `cookieDomain`, `cookieSecure`, etc.
 *
 * @returns A Store object containing the `tokenObject` and optionally a `refresh` object.
 * @throws {AuthKitConfigError} If the `authType` is not supported or if required attributes are missing.
 *
 * @example
 * ```js
 * import createAuthStore from 'react-auth-kit/createAuthStore';
 * const store = createAuthStore('cookie', {
 *   authName: '_auth',
 *   cookieDomain: window.location.hostname,
 *   cookieSecure: window.location.protocol === 'https:',
 *   refresh: createRefresh({
 *    interval: 10,
 *    refreshApiCallback: async (param) => {
 *      ... // Your refresh logic here
 *    };
 *  }
 */
export default function createAuthStore<T>(
  authType: defaultAuthTypes,
  authAttributes: storeAttributes<T>,
): Store<T> {

  // Check if the authType is valid
  // If not, throw an error
  let storage: IStorage;
  if(authType === "cookie"){
    // Check if the cookie attributes are provided
    const cookieSecure = valueOrDefault(authAttributes.cookieSecure, window.location.protocol === 'https:');
    const cookieDomain = valueOrDefault(authAttributes.cookieDomain, window.location.hostname);
    const cookiePath = valueOrDefault(authAttributes.cookiePath, '/');
    const cookieSameSite = valueOrDefault(authAttributes.cookieSameSite, 'none');

    // Create a new CookieStorage instance with the provided attributes
    storage = new CookieStorage(cookieDomain, cookieSecure, cookiePath, cookieSameSite);
  }
  else if (authType === "localstorage"){
    // Create a new LocalStorage instance
    storage = new LocalStorage();
  }
  else {
    throw new AuthKitConfigError(
      `authType '${authType}' is not supported. Use 'cookie' or 'localstorage'`,
    );
  }

  /**
   * Instance the TokenStore with proper parameters
   */
  const tokenStore = new TokenStore<T>(
    !!authAttributes.refresh,
    storage,
    new DefaultStorageNamingStrategy(valueOrDefault(authAttributes.authName, '_auth')),
    new JwtToken(),
    valueOrDefault(authAttributes.debug, false),
  )

  return {
    tokenStore: tokenStore,
    refresh: authAttributes.refresh,
  };
}
