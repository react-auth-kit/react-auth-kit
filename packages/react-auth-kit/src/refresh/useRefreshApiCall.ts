/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type {createRefreshAttribute} from "./createRefresh";
import {ITokenStore} from "../store";
import {RefreshTokenActionResponsePassed} from "./types";

function useRefreshApiCall<T>(refresh: createRefreshAttribute<T>, store: ITokenStore<T>): () => Promise<RefreshTokenActionResponsePassed<T> | null> {
  return async (): Promise<RefreshTokenActionResponsePassed<T> | null> => {
    return refresh.refreshApiCallback({
      refreshToken: store.value.refresh?.token,
    })
      .then((result) => {
        // IF the API call is successful, then refresh the AUTH state
        if (result.isSuccess) {
          // store the new value using the state update
          return new Promise<RefreshTokenActionResponsePassed<T>>((resolve) => {
            resolve(result);
          });
        } else {
          // sign out if failed to refresh
          return new Promise<null>((resolve) => {
            resolve(null);
          });
        }
      })
      .catch((e) => {
        // Retry for Future
        return new Promise<null>((_resolve, reject) => {
          reject(e);
        })
      });
  }
}

export {useRefreshApiCall};
