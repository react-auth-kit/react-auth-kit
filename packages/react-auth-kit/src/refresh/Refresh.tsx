/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {useEffect, useState, type PropsWithChildren, type ReactNode} from "react";
import type {createRefreshAttribute} from "./createRefresh";
import {useRefreshApiCall} from "./useRefreshApiCall";

import {ITokenStore} from "../store";

import {useInterval} from "../utils/hooks";
import Action from "../utils/action";


interface RefreshProps<T> {
  refresh: createRefreshAttribute<T>
  store: ITokenStore<T>
}

/**
 * Refresh Component
 * This component is responsible for handling the refresh logic of the authentication token.
 * It checks if the user is signed in, and if the refresh token is valid, it will refresh the token.
 * If the user is not signed in or the refresh token is expired, it will sign out the user.
 * It also handles the initial refresh state and periodic refresh calls.
 *
 * @typeParam T - Type of the authentication token
 *
 * @param children - Children components to render when the refresh is not in progress
 * @param refresh - The refresh attribute containing the refresh logic and API callback
 * @param store - The token store that holds the authentication state and methods to update it
 *
 * @returns ReactNode - The rendered component, either the children or the initial refresh component
 *
 * @internal
 */
function Refresh<T>({children, refresh, store}: PropsWithChildren<RefreshProps<T>>): ReactNode {
  const [shouldInitialRefreshState, setShouldInitialRefreshState] = useState<boolean>(!!store.value.refresh?.token);
  const refreshApiCall = useRefreshApiCall(refresh, store);

  const _refresh = () => {
    refreshApiCall()
      .then((data) => {
        if (data !== null) {
          Action.doRefresh(data, store);
        } else {
          Action.doSignOut(store);
        }
        setShouldInitialRefreshState(false);
      })
      .catch((e) => {
        console.error(e);
        Action.doSignOut(store);
      });
  }

  useEffect(() => {
    // Periodic call refresh
    useInterval(
      _refresh,
      store.value.isSignIn ? refresh.interval : null,
    );
  }, [store.value.isSignIn]);

  // Initial Refresh
  useEffect(() => {
    // if(!shouldInitialRefreshState){
    //   return;
    // }

    // Check for Initial condition
    const {value} = store;
    // If everything is okay,
    if (value.isSignIn) {
      // Set refreshing to false
      setShouldInitialRefreshState(false);
    }
    // Else If refresh condition met
    else if (value.isUsingRefreshToken && value.refresh && value.refresh.expiresAt > new Date()) {
      // Refresh the auth token by calling the doRefresh
      _refresh()
    }
    // IF the user is sign out
    else {
      // make a SignOut call
      Action.doSignOut(store);
      setShouldInitialRefreshState(false);
    }
    // Set refreshing to false
  }, [])

  if (shouldInitialRefreshState && refresh.initialRefreshComponent) {
    return refresh.initialRefreshComponent;
  } else {
    return children;
  }
}

export {Refresh};
