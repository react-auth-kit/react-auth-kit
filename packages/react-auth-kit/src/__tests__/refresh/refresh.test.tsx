/*
 * Copyright 2025 Arkadip Bhattacharya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// @ts-ignore
import Cookies from "js-cookie";
import {act, render, waitFor} from "@testing-library/react";

import {Refresh, RefreshTokenActionResponse} from "../../refresh";

import createAuthStore from "../../store";
import Action from "../../utils/action";
import createRefresh from "../../refresh/createRefresh";

jest.useFakeTimers()

const doSignOutSpy = jest.spyOn(Action, 'doSignOut');
const doRefreshSpy = jest.spyOn(Action, 'doRefresh');

describe("Initial Refresh", () => {
	const InitialRefreshComponent = () => (
		<div>
			This is the initial Refresh
		</div>
	)

	afterEach(() => {
		Cookies.remove('___auth');
		Cookies.remove('___auth_type');
		Cookies.remove('___state');
		Cookies.remove('___refresh');
		doRefreshSpy.mockClear();
		doSignOutSpy.mockClear();
	});

	it('should not refresh, If all tokens are present', () => {
		Cookies.set('___auth_type', 'Bearer');
    Cookies.set('___state', '{}');
		const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM'+
		'0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw60'+
		'3AjpAqNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8';
		const refreshToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM'+
    '0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw603AjpA'+
    'qNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8';
		Cookies.set('___auth', token);
		Cookies.set('___refresh', refreshToken);


		const apiFn = jest.fn()
		const createRefreshData = createRefresh({
			interval: 10,
			refreshApiCallback: apiFn,
			initialRefreshComponent: <InitialRefreshComponent/>
		});

		const store = createAuthStore("cookie",{
      authName: '__',
      cookieDomain: window.location.hostname,
      cookieSecure: false,
      refresh: createRefreshData
    });

		const {container} = render(
			// @ts-expect-error refresh type
			<Refresh refresh={store.refresh} store={store.tokenStore}>
				<div>
					Container
				</div>
			</Refresh>
		);

		expect(container).toMatchSnapshot();
		expect(apiFn).not.toHaveBeenCalled();
		expect(doSignOutSpy).not.toHaveBeenCalled();
		expect(doRefreshSpy).not.toHaveBeenCalled();
	});

	describe('should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there', () => {
		it('API call is successfully', async () => {
			const refreshToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM'+
			'0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw603AjpA'+
			'qNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8';
			Cookies.set('___refresh', refreshToken);

			const apiFn = jest.fn(
				() => new Promise<RefreshTokenActionResponse<unknown>>((resolve) => {
					setTimeout(resolve, 200, {
						newAuthToken: 'asdf',
						isSuccess: true
					})
				})
			);

			const createRefreshData = createRefresh({
				interval: 10,
				refreshApiCallback: apiFn,
				initialRefreshComponent: <InitialRefreshComponent/>
			});

      const store = createAuthStore("cookie",{
        authName: '__',
        cookieDomain: window.location.hostname,
        cookieSecure: false,
        refresh: createRefreshData,
        debug: false
      });

      expect(store.tokenStore.value).toStrictEqual({
        "auth": null,
        "isSignIn": false,
        "isUsingRefreshToken": true,
        "refresh": {
          "expiresAt": new Date("2223-10-14T04:33:15.000Z"),
          "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw603AjpAqNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8",
        },
        "userState": null,
      });

      const {container} = render(
				// @ts-expect-error refresh type
				<Refresh refresh={store.refresh} store={store.tokenStore}>
					<div>
						Container
					</div>
				</Refresh>
			);

			expect(container).toMatchSnapshot();

      await act(async ()=>{
        await jest.runAllTimersAsync();
      });
			await waitFor(() => expect(apiFn).toHaveBeenCalledTimes(1))

			expect(container).toMatchSnapshot();
			expect(doSignOutSpy).not.toHaveBeenCalled();
			expect(doRefreshSpy).toHaveBeenCalled();
		});
		it('API call is not successfully', async () => {
      const refreshToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM'+
        '0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw603AjpA'+
        'qNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8';
      Cookies.set('___refresh', refreshToken);

      const apiFn = jest.fn(
        () => new Promise<RefreshTokenActionResponse<unknown>>((resolve, reject) => {
          reject("Unknow error occurred")
        })
      );

      const createRefreshData = createRefresh({
        interval: 10,
        refreshApiCallback: apiFn,
        initialRefreshComponent: <InitialRefreshComponent/>
      });

      const store = createAuthStore("cookie",{
        authName: '__',
        cookieDomain: window.location.hostname,
        cookieSecure: false,
        refresh: createRefreshData,
        debug: false
      });

      expect(store.tokenStore.value).toStrictEqual({
        "auth": null,
        "isSignIn": false,
        "isUsingRefreshToken": true,
        "refresh": {
          "expiresAt": new Date("2223-10-14T04:33:15.000Z"),
          "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiZXhwIjo4MDA4NjA1MTk1fQ.ijw603AjpAqNwnUXmv6YB5L6m5aL-llIgBsTJo-k2r8",
        },
        "userState": null,
      });

      const {container} = render(
        // @ts-expect-error refresh type
        <Refresh refresh={store.refresh} store={store.tokenStore}>
          <div>
            Container
          </div>
        </Refresh>
      );

      expect(container).toMatchSnapshot();

      await act(async ()=>{
        await jest.runAllTimersAsync();
      });
      await waitFor(() => expect(apiFn).toHaveBeenCalledTimes(1))

      expect(container).toMatchSnapshot();
      expect(doSignOutSpy).toHaveBeenCalled();
      expect(doRefreshSpy).not.toHaveBeenCalled();
		});
	});

	// describe('If refresh condition is there, and if the refresh component is there, only api call and reducer call and state update', () => {
	// 	it('API call is successfully', () => {

	// 	});
	// 	it('API call is not successfully', () => {

	// 	});
	// });

	it('If no refresh token is there, do not refresh, and call signOut', () => {
		const api = jest.fn()
		const createRefreshData = createRefresh({
			interval: 10,
			refreshApiCallback: api,
			initialRefreshComponent: <InitialRefreshComponent />
		});

    const store = createAuthStore("cookie",{
      authName: '__',
      cookieDomain: window.location.hostname,
      cookieSecure: false,
      refresh: createRefreshData
    });


    const {container} = render(
			// @ts-expect-error refresh type
			<Refresh refresh={store.refresh} store={store.tokenStore}>
				<div>
					Container
				</div>
			</Refresh>
		);

		expect(container).toMatchSnapshot();
		expect(api).not.toHaveBeenCalled();

		expect(doSignOutSpy).toHaveBeenCalled();
		expect(doRefreshSpy).not.toHaveBeenCalled();
	});

	// it('If the refresh component is called once, then it should not call multiple times', () => {

	// });
});

describe('Periodic refresh', () => {
  it('should call API and action, if sign in', () => {

  });
  it('should should calling API and action, after triggering the sign out call', () => {

  });
  it('If sign out, nothing should happened', () => {

	});
});
