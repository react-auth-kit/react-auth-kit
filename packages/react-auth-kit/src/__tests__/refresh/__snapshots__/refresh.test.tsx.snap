// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Refresh If no refresh token is there, do not refresh, and call signOut 1`] = `
<div>
  <div>
    Container
  </div>
</div>
`;

exports[`Initial Refresh should not refresh, If all tokens are present 1`] = `
<div>
  <div>
    Container
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call is not successful 1`] = `
<div>
  <div>
    This is the initial Refresh
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call is not successful 2`] = `
<div>
  <div>
    Container
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call is successful 1`] = `
<div>
  <div>
    This is the initial Refresh
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call is successful 2`] = `
<div>
  <div>
    Container
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call throws an error 1`] = `
<div>
  <div>
    This is the initial Refresh
  </div>
</div>
`;

exports[`Initial Refresh should render initial refresh component and api call and reducer call and state update, If refresh condition is there and if the refresh component is there API call throws an error 2`] = `
<div>
  <div>
    This is the initial Refresh
  </div>
</div>
`;
